File: rukkieapi/routes/services.js
/**
 * GET /v1/services
 *
 * Returns all registered services with their connection status.
 * Called by the RukkiePulse CLI (rukkie scan / rukkie watch).
 *
 * Request headers:
 *   x-rukkie-cli: <CLI_SECRET>
 *
 * Response:
 *   { services: [{ id, name, language, description, activeKeys, lastUsedAt }] }
 */

const express = require('express');
const supabase = require('../db');

const router = express.Router();

const CLI_SECRET = process.env.RUKKIE_CLI_SECRET || 'rukkie-cli-v1-xqmjdjjwprnqogokoejz';

router.get('/', async (req, res) => {
  const cliHeader = req.headers['x-rukkie-cli'] ?? '';
  if (cliHeader !== CLI_SECRET) {
    return res.status(401).json({ error: 'Unauthorized' });
  }

  const { data: services, error } = await supabase
    .from('services')
    .select('id, name, language, description, created_at, api_keys(id, label, key_prefix, last_used_at, revoked_at)')
    .order('created_at', { ascending: true });

  if (error) {
    return res.status(500).json({ error: error.message });
  }

  const result = (services ?? []).map((svc) => {
    const keys = Array.isArray(svc.api_keys) ? svc.api_keys : [];
    const activeKeys = keys.filter((k) => !k.revoked_at);

    let lastUsedAt = null;
    for (const k of activeKeys) {
      if (!lastUsedAt || (k.last_used_at && k.last_used_at > lastUsedAt)) {
        lastUsedAt = k.last_used_at;
      }
    }

    return {
      id: svc.id,
      name: svc.name,
      language: svc.language ?? 'other',
      description: svc.description ?? '',
      activeKeys: activeKeys.length,
      lastUsedAt,
    };
  });

  return res.json({ services: result });
});

module.exports = router;

File: internal/cloud/services.go
package cloud

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

const (
	supabaseURL = "https://xqmjdjjwprnqogokoejz.supabase.co"
	cliSecret   = "rukkie-cli-v1-xqmjdjjwprnqogokoejz"
)

// ServiceStatus holds the cloud status for a registered service.
type ServiceStatus struct {
	ID          string  `json:"id"`
	Name        string  `json:"name"`
	Language    string  `json:"language"`
	Description string  `json:"description"`
	ActiveKeys  int     `json:"activeKeys"`
	LastUsedAt  *string `json:"lastUsedAt"`
}

type servicesResponse struct {
	Services []ServiceStatus `json:"services"`
	Error    string          `json:"error"`
}

// FetchServices calls the services-status Edge Function and returns all registered services.
func FetchServices() ([]ServiceStatus, error) {
	url := supabaseURL + "/functions/v1/services-status"

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("x-rukkie-cli", cliSecret)

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to reach RukkiePulse cloud: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	var result servicesResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("invalid response from cloud: %w", err)
	}

	if result.Error != "" {
		return nil, fmt.Errorf("cloud error: %s", result.Error)
	}

	return result.Services, nil
}

// ConnectionStatus returns the emoji dot, label, and last-seen text for a service.
func ConnectionStatus(lastUsedAt *string) (dot, label, lastSeen string) {
	if lastUsedAt == nil || *lastUsedAt == "" {
		return "⚫", "Never connected", ""
	}

	t, err := time.Parse(time.RFC3339, *lastUsedAt)
	if err != nil {
		return "⚫", "Never connected", ""
	}

	ago := time.Since(t)
	mins := ago.Minutes()

	switch {
	case mins < 5:
		return "🟢", "Live", fmt.Sprintf("last seen %dm ago", int(mins))
	case mins < 60:
		return "🟡", "Recent", fmt.Sprintf("last seen %dm ago", int(mins))
	default:
		return "🔴", "Inactive", fmt.Sprintf("last seen %s", t.Format("2006-01-02 15:04"))
	}
}
